In [ ]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

In [1]:
import pandas as pd
df = pd.read_csv("../input/house-prices/house_prices.csv", na_values = "?")
label = "SalePrice"
X, y = df.drop(columns = label), df[label]
y = (y > 200_000).astype(int)

In [2]:
X

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence  \
0            Lvl    AllPub  ...           0        0    NaN    NaN   
1            Lvl    AllPub  ...           0        0    NaN    NaN   
2            Lvl    AllPub  ...           0        0    NaN    NaN   
3            Lvl    AllPub  ...           0        0    NaN    NaN   
4            Lvl    AllPub  ...           0        0    NaN    NaN   
...          ...       ...  ...         ...      ...    ...    ...   
1455         Lvl    AllPub  ...           0        0    NaN    NaN   
1456         Lvl    AllPub  ...           0        0    NaN  MnPrv   
1457         Lvl    AllPub  ...           0        0    NaN  GdPrv   
1458         Lvl    AllPub  ...           0        0    NaN    NaN   
1459         Lvl    AllPub  ...           0        0    NaN    NaN   

     MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
0            NaN       0      2    2008        WD         Normal  
1            NaN       0      5    2007        WD         Normal  
2            NaN       0      9    2008        WD         Normal  
3            NaN       0      2    2006        WD        Abnorml  
4            NaN       0     12    2008        WD         Normal  
...          ...     ...    ...     ...       ...            ...  
1455         NaN       0      8    2007        WD         Normal  
1456         NaN       0      2    2010        WD         Normal  
1457        Shed    2500      5    2010        WD         Normal  
1458         NaN       0      4    2010        WD         Normal  
1459         NaN       0      6    2008        WD         Normal  

[1460 rows x 80 columns]

In [4]:
num_features = ['LotArea', 'FullBath', 'HalfBath']
cat_features = ['Neighborhood', 'HouseStyle']
X = X[num_features + cat_features]

In [6]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder

num_transformer = Pipeline(steps = [
    ("imputer", SimpleImputer(strategy = "median")),
    ("scaler", StandardScaler()),
])

cat_transformer = OneHotEncoder(handle_unknown = "ignore")

In [9]:
# Apply the transformations using "ColumnTransformer"
from sklearn.compose import ColumnTransformer

preprocessor = ColumnTransformer(transformers=[
    ('num', num_transformer, num_features),
    ('cat', cat_transformer, cat_features),
])

In [14]:
from sklearn.linear_model import LogisticRegression

model = Pipeline(steps = [
    ("preprocessor", preprocessor),
    ("classifier", LogisticRegression())
    ]
)

In [15]:
from sklearn import set_config

set_config(display = "diagram")
model

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='median')),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['LotArea', 'FullBath',
                                                   'HalfBath']),
                                                 ('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['Neighborhood',
                                                   'HouseStyle'])])),
                ('classifier', LogisticRegression())])

In [16]:
from sklearn.model_selection import cross_validate

cv_results = cross_validate(model, X, y, cv=5)
scores = cv_results["test_score"]
print("The mean cross-validation accuracy is: "
      f"{scores.mean():.3f} +/- {scores.std():.3f}")

The mean cross-validation accuracy is: 0.859 +/- 0.018


<div class="admonition note alert alert-info">
<p class="first admonition-title" style="font-weight: bold;">Note</p>
<p>In this case, around 86% of the times the pipeline correctly predicts whether
the price of a house is above or below the 200_000 dollars threshold. But
be aware that this score was obtained by picking some features by hand, which
is not necessarily the best thing we can do for this classification task. In this
example we can hope that fitting a complex machine learning pipelines on a
richer set of features can improve upon this performance level.</p>
<p class="last">Reducing a price estimation problem to a binary classification problem with a
single threshold at 200_000 dollars is probably too coarse to be useful in
in practice. Treating this problem as a regression problem is probably a better
idea. We will see later in this MOOC how to train and evaluate the performance
of various regression models.</p>
</div>